pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -137,23 +138,59 @@ func (configgen *ConfigGeneratorImpl) BuildDeltaClusters(proxy *model.Proxy, pus
 	envoyFilterPatches := push.EnvoyFilters(proxy)
 	cb := NewClusterBuilder(proxy, updates, configgen.Cache)
 	instances := proxy.ServiceInstances
+
 	for key := range updates.ConfigsUpdated {
 		// get service that changed
-		service := push.ServiceForHostname(proxy, host.Name(key.Name))
-		// SidecarScope.Service will return nil if the proxy doesn't care about the service OR it was deleted.
-		// we can cross reference with WatchedResources to figure out which services were deleted.
-		if service == nil {
-			// WatchedResources.ResourceNames will contain the names of the clusters it is subscribed to. We can
-			// check with the name of our service (cluster names are in the format outbound|<port>||<hostname>
-			// so, we can check if the cluster names contains the service name, and determine if it is deleted by that
-			for _, n := range watched.ResourceNames {
-				_, _, svcHost, _ := model.ParseSubsetKey(n)
-				if svcHost == host.Name(key.Name) {
-					removedClusterNames = append(removedClusterNames, n)
-				}
+		if key.Kind == gvk.ServiceEntry {
+			service := push.ServiceForHostname(proxy, host.Name(key.Name))
+			// SidecarScope.ServiceForHostname will return nil if the proxy doesn't care about the service OR it was deleted.
+			// we can cross-reference with WatchedResources to figure out which services were deleted.
+			if service == nil {
+				// we assume a service was deleted, it doesn't hurt to have positives for removedResources when a resource included
+				// does not exist.
+				// WatchedResources.ResourceNames will contain the names of the clusters it is subscribed to. We can
+				// check with the name of our service (cluster names are in the format outbound|<port>||<hostname>
+				// so, we can check if the cluster names contains the service name, and determine if it is deleted by that
+				removedClusterNames = append(removedClusterNames, getRemovedClusterNames(watched.ResourceNames, nil, key.Name)...)
+			} else {
+				services = append(services, service)
 			}
 		} else {
-			services = append(services, service)
+			// destination rules
```

Comment:
It makes the implementation complicated, someday we want to add other kinds, we need to add all of them.

Reply:
This makes me doubt whether this is a good way to do delta xds, @howardjohn

Reply:
I generally agree this is pretty complicated but I am not sure of better options... any ideas?

Reply:
All I can think of is a large switch/case or something, breaking it out into further abstraction than a separate method seems like overkill

Reply:
What are other possible different config types that come here for cluster?